toolkit/tools/internal/retry/retry.go

```diff
@@ -40,21 +44,25 @@ func calculateLinearDelay(failCount int, sleep time.Duration) time.Duration {
 	return sleep * time.Duration(failCount)
 }
 
-// backoffSleep sleeps for the given duration, unless the cancel channel is triggered or closed first. The cancel channel
-// may be nil in which case the sleep will always complete.
-func backoffSleep(delay time.Duration, cancel <-chan struct{}) (cancelled bool) {
+// backoffSleep sleeps for the given duration, unless the context is cancelled. The context
+// may be nil in which case the background context will be used and the sleep will always complete.
+func backoffSleep(ctx context.Context, delay time.Duration) (cancelled bool, err error) {
+	if ctx == nil {
+		return false, ErrNilContext
+	}
+
 	// Check if we were cancelled before we sleep to avoid a race condition with
 	// delay=0 and an immediate cancel.
 	select {
```

Comment:
So if I'm understanding this right, the cancellation function will cause ctx.Done() to start reporting true? 

So at any point if any other section of the code calls the cancaelFunc, next time backoffSleep is called it will propagate up the chain ending loops and gracefully exiting the context?

Reply:
Correct (although to be accurate `ctx.Done()` is a `chan`, and reading from a closed channel in a `select` block allows that branch to be executed). The idea of a context is it handles all the synchronization of canceling a job for you.